src/cryptsetup/cryptsetup.c

```diff
@@ -2265,6 +2269,44 @@ static void remove_and_erasep(const char **p) {
                 log_warning_errno(r, "Unable to erase key file '%s', ignoring: %m", *p);
 }
 
+static const char* determine_token_type(void) {
+        if (arg_tpm2_device || arg_tpm2_device_auto)
+                return "tpm2";
+        else if (arg_fido2_device || arg_fido2_device_auto)
+                return "fido2";
+        else if (arg_pkcs11_uri || arg_pkcs11_uri_auto)
+                return "pkcs11";
+        return NULL;
+}
```

Comment:
so, hmm, looking at this I kinda get the feeling we should just go for a proper enum here, like we usually do, and pass that around. i.e.:

```c
typedef enum TokenType {
        TOKEN_TPM2,
        TOKEN_FIDO2,
        TOKEN_PKCS11,
        _TOKEN_TYPE_MAX,
        _TOKEN_TYPE_INVALID = -EINVAL,
} TokenType;

static const char* const token_type_table[_TOKEN_TYPE_MAX] = {
        [TOKEN_TPM2]   = "tpm2",
        [TOKEN_FIDO2]  = "fido2",
        [TOKEN_PKCS11] = "pkcs11",
};

DEFINE_PRIVATE_STRING_TABLE_LOOKUP_TO_STRING(token_type, TokenType);

static TokenType determine_token_type(void) {
        if (arg_tpm2_device || arg_tpm2_device_auto)
                return TOKEN_TPM2;
        … 
        return _TOKEN_TYPE_INVALID;   
}
```

And thwn we'd pass around TokenType everywhere, and when we need it as string we just do `token_type_to_string(token_type)`. And if we want to check if it's not passed we'd just do `token_type < 0`.

That would be more along how we do these things otherwise, and I think it would be more feature proof too.